File: apps/android/app/src/main/java/com/teslasync/widget-primitives/WidgetShell/WidgetShellModel.kt
// Pure, framework-free model + projection + diagnostics for the WidgetShell widget primitive — the native
// analogue of every layout decision the web source makes (web/src/features/dashboard/widgets/WidgetShell.tsx)
// before Compose paints anything. No Compose, no Android framework, no HTTP: every declaration here is exercised
// off-device in the :android:testReleaseUnitTest gate, so the composable stays a thin render layer over these
// pure functions (the accepted sibling-surface contract — WidgetBigNumber / WidgetComparisonCard models).
//
// What the web source IS (and therefore the complete behaviour this primitive reproduces): the shared chrome that
// wraps every dashboard widget. It is purely presentational — it performs no query and owns no data of its own;
// every visible string (title, help text, actions) is handed in by the host. It plays three top-level branches in
// strict precedence:
//   1. `loading`  → a full-cell `Skeleton` (early return, no header, no children).
//   2. `error`    → a centered `QueryError` (early return). The web wraps the message in `new Error(error)`, which
//                   carries no HTTP status, so the web `QueryError` falls through to its generic "can't reach
//                   server" (network) branch and shows CANNED recovery copy — never the raw message. The `error`
//                   string is therefore a PRESENCE signal, not display text (honesty covenant: no silent drift —
//                   we reproduce exactly what the web renders, which is the classified panel, not the message).
//   3. content    → the header (icon + uppercase muted title + optional help "?" + freshness chip + optional pin +
//                   actions) over the body slot. When there is no title the header collapses to an overlay
//                   freshness chip (top-end) plus an optional right-aligned actions row, exactly as the web `title
//                   ? … : …` ternary does.
//
// Freshness: the web exposes two mutually-exclusive input modes — granular props (`updatedAt`/`isFetching`/
// `isStale`/`isError`) or an entire TanStack query (`query`). Granular wins when `updatedAt` is supplied; otherwise
// the query is unwrapped (the web `DataFreshnessAuto` does literally this: `updatedAt = dataUpdatedAt > 0 ?
// dataUpdatedAt : null`, then the same four flags). [WidgetShellModel.project] normalizes BOTH modes into one
// [WidgetShellFreshnessState] so the composable renders a single shared `DataFreshness` atom — the four data-surface
// states the prompt enumerates (fresh / fetching / stale / error→offline) are reproduced THROUGH that atom, with
// `compact = !title` mirroring the web `freshnessCompact`.
//
// Why "empty" is not a branch here: like the web shell, this primitive delegates the empty state to its `content`
// slot — the host passes an `EmptyState` as the body when it has no rows (honesty covenant: no scope narrowing —
// the empty surface is reproduced, just owned by the caller exactly as on the web). The "offline" state is the
// freshness atom's `isError` tier (web maps the error freshness to the offline chip), so it too is reproduced.
//
// `InvalidPackageDeclaration` is suppressed because this primitive's mandated directory
// (com/teslasync/widget-primitives/WidgetShell — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling widget / shared surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.widgetprimitives.widgetshell

import io.teslasync.shared.core.diagnostics.Logger

/**
 * Canonical registry metadata for the WidgetShell primitive. The diagnostics [SLUG] is emitted with the one-shot
 * `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`WidgetShell`).
 */
object WidgetShellRegistration {
    /** Stable primitive id (also the key a host would bind the primitive with). */
    const val ID: String = "widget-shell"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11, prompt-mandated). */
    const val SLUG: String = "WidgetShell"
}

/**
 * The web-default knobs kept as named constants so the composable and the unit gate agree on one source of truth —
 * no loose numerals drift between the render layer and its tests.
 */
object WidgetShellDefaults {
    /**
     * The green "just updated" glow hold in milliseconds — the web `setTimeout(…, 1500)` that clears the
     * `shadow-[…]` pulse 1.5s after the data timestamp changes.
     */
    const val PULSE_HOLD_MS: Long = 1_500L
}

/**
 * The three top-level branches the web shell plays, in strict precedence (loading wins over error wins over
 * content), so the composable renders exactly one and never paints children during loading/error.
 */
enum class WidgetShellPhase { Loading, Error, Content }

/**
 * The TanStack-query freshness input — the native analogue of the web `FreshnessQuery`
 * (`Pick<UseQueryResult, 'isFetching' | 'isStale' | 'isError' | 'dataUpdatedAt' | 'refetch'>`). The `refetch`
 * member is supplied separately as the composable's `onRefresh` lambda (a render concern), so this pure type
 * carries only the four data flags.
 *
 * @param dataUpdatedAtMillis ms timestamp of the last successful fetch (0 = never), web `dataUpdatedAt`.
 * @param isFetching whether a fetch is in flight (web `isFetching`).
 * @param isStale whether the data is past its staleTime (web `isStale`).
 * @param isError whether the query is in an error state (web `isError`).
 */
data class WidgetShellFreshnessQuery(
    val dataUpdatedAtMillis: Long,
    val isFetching: Boolean = false,
    val isStale: Boolean = false,
    val isError: Boolean = false,
)

/**
 * Optional contextual help for the widget title — the native analogue of the web `help?: WidgetHelp`. Native i18n
 * resolves at the call site (`stringResource`), so the host hands in the already-localized [text]; the shell
 * forwards it to the `HelpIcon` atom together with a title-derived screen-reader label (the web "More info about
 * {title}" `ariaLabel`). The web `WidgetHelp.learnMore` is a capability of the out-of-scope HelpTooltip atom (the
 * P3 component-library bundle), not of this shell, so it is not part of the primitive's surface.
 *
 * @param text the resolved help body shown in the tooltip (web `help.text` / resolved `help.i18nKey`).
 */
data class WidgetShellHelp(
    val text: String,
)

/**
 * The normalized freshness the composable paints — the single shape both web input modes (granular props and the
 * `query`) collapse into, so one `DataFreshness` atom serves every caller.
 *
 * @param updatedAtMillis the last-fetch timestamp, or `null` when never fetched (web `updatedAt > 0 ? … : null`).
 * @param isFetching whether a fetch is in flight (animated chip).
 * @param isStale whether the data is past its staleTime (amber chip).
 * @param isError whether the fetch failed (the web error→offline chip).
 * @param compact dot-only chip with no relative-time text — the web `freshnessCompact = !title`.
 */
data class WidgetShellFreshnessState(
    val updatedAtMillis: Long?,
    val isFetching: Boolean,
    val isStale: Boolean,
    val isError: Boolean,
    val compact: Boolean,
)

/**
 * The presentational inputs the pure [WidgetShellModel.project] reduces — the native analogue of the web
 * `WidgetShellProps` minus the render-only slots (`icon`/`actions`/`children` Compose lambdas) and the `onRefresh`
 * callback, which live on the composable. Grouped into one spec so the projection stays under the parameter budget
 * and the composable has a single value to derive from.
 *
 * @param title the widget title; blank collapses to no header (web `title`).
 * @param loading whether the first fetch is in flight (web `loading`).
 * @param error the error PRESENCE signal; non-empty triggers the error branch (web `error`).
 * @param updatedAtMillis granular last-fetch timestamp, or `null` when the caller uses the [query] mode instead
 *   (web `updatedAt` — `undefined` ⇒ not the granular mode).
 * @param isFetching granular fetching flag (web `isFetching`, default `false`).
 * @param isStale granular stale flag (web `isStale`, default `false`).
 * @param isError granular error flag (web `isError`, default `false`).
 * @param query the TanStack-query freshness mode, used when [updatedAtMillis] is `null` (web `query`).
 * @param hasHelp whether contextual help was supplied (web `help != null`); the "?" shows only WITH a title.
 * @param widgetId the stable widget id; with [dashboardId] it gates the pin affordance (web `widgetId`).
 * @param dashboardId the per-dashboard pin context (web `dashboardId`).
 */
data class WidgetShellSpec(
    val title: String? = null,
    val loading: Boolean = false,
    val error: String? = null,
    val updatedAtMillis: Long? = null,
    val isFetching: Boolean = false,
    val isStale: Boolean = false,
    val isError: Boolean = false,
    val query: WidgetShellFreshnessQuery? = null,
    val hasHelp: Boolean = false,
    val widgetId: String? = null,
    val dashboardId: String? = null,
)

/**
 * The render-ready projection of a [WidgetShellSpec] — the pure data the composable paints. The title is
 * normalized (blank collapses to `null` so a stray empty prop never reserves the header), the freshness is
 * resolved from whichever input mode the caller used, and the pin/help gates are pre-computed so the render layer
 * is a flat `when`.
 *
 * @param phase the single branch to render (web's `loading` / `error` early returns vs the content body).
 * @param title the normalized title, or `null` for a title-less (overlay-freshness) widget.
 * @param showHelp whether the help "?" renders — web `help && title` (help shows only alongside a title).
 * @param freshness the normalized freshness chip, or `null` when the caller supplied neither mode.
 * @param showPin whether the pin renders — web `widgetId && dashboardId`.
 * @param effectiveUpdatedAtMillis the timestamp the pulse-on-change effect watches — web
 *   `updatedAt ?? query?.dataUpdatedAt` (the RAW value, before the `> 0` normalization).
 */
data class WidgetShellContent(
    val phase: WidgetShellPhase,
    val title: String?,
    val showHelp: Boolean,
    val freshness: WidgetShellFreshnessState?,
    val showPin: Boolean,
    val effectiveUpdatedAtMillis: Long?,
)

/**
 * The pure projection the composable renders — a 1:1 port of the layout decisions the web `WidgetShell` makes.
 * Stateless and side-effect-free so it is fully covered by the off-device unit gate; the composable only drives
 * the pulse animation clock, resolves i18n strings, and slots the host's `icon`/`actions`/`children`.
 */
object WidgetShellModel {
    /**
     * Reduce a [spec] to its render-ready [WidgetShellContent].
     *
     * Precedence mirrors the web early-returns exactly: `loading` wins over a non-empty `error`, which wins over
     * the content body. The title is trimmed and blank-normalized to `null`. Freshness resolves from the granular
     * props when [WidgetShellSpec.updatedAtMillis] is supplied, otherwise from [WidgetShellSpec.query], otherwise
     * to `null`; either way the timestamp is `> 0`-normalized so a never-fetched widget reads "never updated"
     * rather than the epoch. `compact` follows the web `!title`. The pin and help gates match the web truthiness
     * checks (`widgetId && dashboardId`, `help && title`).
     */
    fun project(spec: WidgetShellSpec): WidgetShellContent {
        val title = spec.title?.trim()?.takeIf { it.isNotEmpty() }
        val phase =
            when {
                spec.loading -> WidgetShellPhase.Loading
                // Web `if (error)` — JS string truthiness: "" is falsy, any non-empty string (incl. whitespace) is
                // truthy. isNullOrEmpty reproduces that exactly.
                !spec.error.isNullOrEmpty() -> WidgetShellPhase.Error
                else -> WidgetShellPhase.Content
            }

        val freshness = resolveFreshness(spec, compact = title == null)

        return WidgetShellContent(
            phase = phase,
            title = title,
            showHelp = spec.hasHelp && title != null,
            freshness = freshness,
            showPin = !spec.widgetId.isNullOrEmpty() && !spec.dashboardId.isNullOrEmpty(),
            effectiveUpdatedAtMillis = spec.updatedAtMillis ?: spec.query?.dataUpdatedAtMillis,
        )
    }

    /**
     * Resolve the single normalized freshness chip from whichever input mode the caller used. Granular props win
     * when [WidgetShellSpec.updatedAtMillis] is non-null (web `updatedAt !== undefined`); otherwise the
     * [WidgetShellSpec.query] is unwrapped (the web `DataFreshnessAuto` derivation); otherwise there is no chip.
     */
    private fun resolveFreshness(
        spec: WidgetShellSpec,
        compact: Boolean,
    ): WidgetShellFreshnessState? =
        when {
            spec.updatedAtMillis != null ->
                WidgetShellFreshnessState(
                    updatedAtMillis = spec.updatedAtMillis.takeIf { it > 0 },
                    isFetching = spec.isFetching,
                    isStale = spec.isStale,
                    isError = spec.isError,
                    compact = compact,
                )

            spec.query != null ->
                WidgetShellFreshnessState(
                    updatedAtMillis = spec.query.dataUpdatedAtMillis.takeIf { it > 0 },
                    isFetching = spec.query.isFetching,
                    isStale = spec.query.isStale,
                    isError = spec.query.isError,
                    compact = compact,
                )

            else -> null
        }

    /**
     * Whether the green "just updated" glow should fire — the web pulse effect condition. Pulses only when the
     * effective timestamp moved to a NEW positive value AND a previous value was already observed (the web
     * `prevUpdatedAt.current !== undefined`), so the very first observation never flashes.
     *
     * @param previousMillis the last observed effective timestamp, or `null` when none has been seen yet.
     * @param currentMillis the current effective timestamp.
     */
    fun shouldPulse(
        previousMillis: Long?,
        currentMillis: Long?,
    ): Boolean =
        currentMillis != null &&
            currentMillis > 0 &&
            previousMillis != null &&
            previousMillis != currentMillis
}

/**
 * PII-safe diagnostics for the primitive (P1/S11). Emits only the stable, dot-namespaced `view.opened` event
 * tagged with the surface [WidgetShellRegistration.SLUG] — never the title, help text, freshness timestamp, or any
 * host-supplied copy — so a diagnostics line can never leak what the shell wraps. Kept free of Compose so it is
 * unit-tested with a recording [Logger]; the composable calls it once per open.
 */
object WidgetShellDiagnostics {
    private const val VIEW_OPENED: String = "view.opened"
    private const val SURFACE_KEY: String = "surface"

    /** Emits the one PII-safe `view.opened` diagnostic. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to WidgetShellRegistration.SLUG))
    }
}

File: apps/apple/TeslaSync/widget-primitives/WidgetShell.Model.swift
//
//  WidgetShell.Model.swift
//  TeslaSync — P4 widget primitive · 0013 · WidgetShell (Apple)
//
//  Pure (SwiftUI-free) model, render-state decisions, freshness logic, accessibility, i18n facade
//  and telemetry seam for the WidgetShell primitive — the native parity of
//  features/dashboard/widgets/WidgetShell.tsx. Keeping these Foundation-only lets the projection /
//  decision logic compile and unit-test on a plain host; the SwiftUI chrome layers on top in
//  WidgetShell.swift (+ subviews in WidgetShell.Views.swift).
//
//  WidgetShell is the shared widget-chrome building block: it wraps a caller-supplied content slot
//  with an optional titled header (icon + title + contextual "?" help + data-freshness chip + pin +
//  actions), renders the loading skeleton and the query-error state, and pulses softly when the data
//  timestamp changes. It owns no networking and no data source — every input is supplied by the
//  hosting widget (web: a pure presentational component).
//

import Foundation
import OSLog

// MARK: - Contextual help metadata (web `WidgetHelp` from widgets/types.ts)

/// A "Learn more" link rendered under the help body — the native port of the web
/// `learnMore?: { url; label? }`.
public struct WidgetHelpLink: Equatable, Sendable {
    public let url: URL
    public let label: String?

    public init(url: URL, label: String? = nil) {
        self.url = url
        self.label = label
    }
}

/// Metadata describing a widget's contextual help — the native port of the web `WidgetHelp`
/// (`{ text?, i18nKey?, defaultValue?, learnMore? }`). Resolution mirrors the web `HelpTooltip`:
/// prefer the translated `i18nKey` (with `defaultValue` fallback), else the literal `text`; an empty
/// result yields `nil` so the host can omit the affordance entirely (web `if (!resolved) return null`).
public struct WidgetHelp: Equatable, Sendable {
    public let text: String?
    public let i18nKey: String?
    public let defaultValue: String?
    public let learnMore: WidgetHelpLink?

    public init(
        text: String? = nil,
        i18nKey: String? = nil,
        defaultValue: String? = nil,
        learnMore: WidgetHelpLink? = nil
    ) {
        self.text = text
        self.i18nKey = i18nKey
        self.defaultValue = defaultValue
        self.learnMore = learnMore
    }

    /// Web `resolved = i18nKey ? t(i18nKey, defaultValue ?? '') : (text ?? '')`, returning `nil` for
    /// the empty result so the help button is omitted. `localize` is the P1/S10 facade
    /// `(key, fallback) -> String`, injected so this stays SwiftUI-free and unit-testable.
    public func resolvedText(localize: (_ key: String, _ fallback: String) -> String) -> String? {
        let resolved: String = if let i18nKey, !i18nKey.isEmpty {
            localize(i18nKey, defaultValue ?? "")
        } else {
            text ?? ""
        }
        return resolved.isEmpty ? nil : resolved
    }
}

// MARK: - Render state (web `if (loading) … ; if (error) …`)

/// The top-level render branch the shell resolves before drawing chrome, mirroring the web early
/// returns: `loading` wins over `error`, otherwise the titled/title-less content surface renders.
public enum WidgetShellState: Equatable, Sendable {
    case loading
    case error
    case ready

    /// Web precedence: `if (loading) return <Skeleton/>; if (error) return <QueryError/>;`. An empty
    /// error string is falsy in JS, so it is treated as "no error" (→ `ready`).
    public static func resolve(loading: Bool, error: String?) -> WidgetShellState {
        if loading { return .loading }
        if let error, !error.isEmpty { return .error }
        return .ready
    }
}

// MARK: - Data-freshness status (web `DataFreshness` four-state machine)

/// The freshness chip's status, mapped from a TanStack-Query-like result exactly as the web
/// `DataFreshness` does: `isError → error`, else `isFetching → fetching`, else `isStale → stale`,
/// else `fresh`.
public enum WidgetShellFreshnessStatus: String, Equatable, Sendable, CaseIterable {
    case fresh
    case fetching
    case stale
    case error

    public static func resolve(isError: Bool, isFetching: Bool, isStale: Bool) -> WidgetShellFreshnessStatus {
        if isError { return .error }
        if isFetching { return .fetching }
        if isStale { return .stale }
        return .fresh
    }

    /// P1/S10 key for the spoken status word (web embeds the raw status in `a11y.dataFreshness`).
    public var localizationKey: String {
        "freshness.status.\(rawValue)"
    }
}

// MARK: - Relative-time bucket (web `formatRelativeTime`)

/// A coarse "time since update" bucket — the SwiftUI-free core of the web `formatRelativeTime`
/// (`<60s → just now`, `<60m → Nm`, `<24h → Nh`, `<7d → Nd`, else `Nw`). The view resolves the bucket
/// to a localized string at the display boundary (P1/S10), keeping the threshold logic testable.
public enum WidgetShellRelativeTimeBucket: Equatable, Sendable {
    case justNow
    case minutes(Int)
    case hours(Int)
    case days(Int)
    case weeks(Int)

    public static func bucket(updatedAtMillis: Double, nowMillis: Double) -> WidgetShellRelativeTimeBucket {
        // Web `Math.floor((Date.now() - ms) / 1000)`.
        let seconds = Int(floor((nowMillis - updatedAtMillis) / 1000.0))
        if seconds < 60 { return .justNow }
        if seconds < 3600 { return .minutes(seconds / 60) }
        if seconds < 86400 { return .hours(seconds / 3600) }
        if seconds < 604_800 { return .days(seconds / 86400) }
        return .weeks(seconds / 604_800)
    }
}

// MARK: - Freshness label composition (web `relativeTime` ternary)

/// What the freshness chip shows after its dot + icon — the native port of the web
/// `relativeTime = updatedAt && !isFetching ? formatRelative : isFetching ? 'updating…' :
/// isError ? 'error' : ''`. `none` renders no trailing text.
public enum WidgetShellFreshnessLabel: Equatable, Sendable {
    case relative(WidgetShellRelativeTimeBucket)
    case updating
    case error
    case none

    public static func resolve(
        updatedAtMillis: Double?,
        isFetching: Bool,
        isError: Bool,
        nowMillis: Double
    ) -> WidgetShellFreshnessLabel {
        if let updatedAtMillis, updatedAtMillis > 0, !isFetching {
            return .relative(.bucket(updatedAtMillis: updatedAtMillis, nowMillis: nowMillis))
        }
        if isFetching { return .updating }
        if isError { return .error }
        return .none
    }
}

// MARK: - Pulse-on-update (web `justUpdated` effect)

/// The decision behind the web `justUpdated` 1.5 s green-glow effect: pulse only when a *prior*
/// timestamp existed (not the first value) and the new, positive timestamp differs from it
/// (web `prevUpdatedAt.current !== undefined && prevUpdatedAt.current !== effectiveUpdatedAt`).
public enum WidgetShellPulse {
    public static func shouldPulse(previous: Double?, next: Double?) -> Bool {
        guard let next, next > 0 else { return false }
        guard let previous else { return false }
        return previous != next
    }
}

// MARK: - Header / freshness layout decisions (web `title ? … : …`, `freshnessCompact = !title`)

/// Pure layout decisions mirroring the web component's conditional structure so they can be unit
/// tested without SwiftUI.
public enum WidgetShellLayout {
    /// Web renders the titled header only for a truthy `title`; an empty/absent title falls through
    /// to the title-less branch (overlay freshness).
    public static func showsTitleHeader(title: String?) -> Bool {
        guard let title else { return false }
        return !title.isEmpty
    }

    /// Web `freshnessCompact = !title` — title-less widgets (typically 1×1) show the dot-only chip.
    public static func freshnessIsCompact(title: String?) -> Bool {
        !showsTitleHeader(title: title)
    }
}

// MARK: - Accessibility label builders (pure, testable)

/// VoiceOver label composition for the shell's interactive chrome. Format strings come from the
/// P1/S10 facade; these helpers stay pure so the join logic is unit-testable.
public enum WidgetShellAccessibility {
    /// Web `aria-label={'Data freshness: {{state}}'}` for the non-interactive freshness chip.
    public static func dataFreshnessLabel(format: String, status: String) -> String {
        String(format: format, status)
    }

    /// Web `ariaLabel={`More info about ${title}`}` for the help trigger.
    public static func helpLabel(format: String, title: String) -> String {
        String(format: format, title)
    }
}

// MARK: - Diagnostics surface slug (P1/S11 `view.opened`)

/// Stable, non-generic home for the surface slug so tests can reference it without spelling out the
/// view's generic parameters.
public enum WidgetShellSurface {
    public static let slug = "WidgetShell"
}

// MARK: - Telemetry seam (P1/S11 diagnostics contract)

/// Emits the `view.opened` diagnostics event for a surface. The default implementation logs via
/// `os.Logger`; the production app injects an adapter forwarding to the shared core telemetry
/// (consent-gated + redacted there). Only the stable surface slug is emitted — never PII.
public protocol WidgetShellTelemetry: Sendable {
    func viewOpened(surface: String)
}

/// `os.Logger`-backed default that records the surface open as a redaction-safe `view.opened` event
/// carrying only the public surface slug.
public struct OSLogWidgetShellTelemetry: WidgetShellTelemetry {
    private let logger: Logger

    public init() {
        logger = Logger(subsystem: "io.teslasync.app", category: "diagnostics")
    }

    public func viewOpened(surface: String) {
        logger.info("view.opened surface=\(surface, privacy: .public)")
    }
}

// MARK: - Localization facade (P1/S10) — web `t(key, default)`

/// Resolves the surface's strings by key with the web English fallback, so the view holds no
/// hardcoded literals. Keys live in the "WidgetShell" table, folded into the app
/// `Localizable.xcstrings` catalog at integration time. The web `WidgetShell` is anonymous; the
/// strings it renders come from the chrome it composes (`DataFreshness`, `HelpTooltip`, `PinButton`,
/// `QueryError`), captured verbatim here.
public enum WidgetShellStrings {
    public static let table = "WidgetShell"

    public static func string(_ key: String, _ fallback: String) -> String {
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }
}
